nixos: rebuild less derivations on nixpkgs/config hash changes.

r-vdp opened the pull request:
###### Description of changes

Before this change, nixpkgs and configuration revision hashes were hard-coded into the `nixos-version` script.
This means that even a trivial change (like an empty commit), would trigger a rebuild of every derivation that depends on the system path and every derivation that depends on those. On my config, this meant a minimum of 42 derivations that had to be rebuilt for every change, which adds to the feedback loop when hacking on my config.

With the change of this PR, `nixos-version` now reads a JSON file from /run/current-system instead of hard-coding the values in the script, which means that the `system-path` derivation no longer changes every time one of the commit revisions changes.  For my config, this change reduced the amount of derivations that need to be rebuilt after a trivial config change from 42 to 9.

On top of this, I restructured the tools module a bit so that it becomes possible to override the packages for these tools if needed, which was not possible before.

EDIT: as suggested by reviewers, I made the new behaviour opt-in. We can re-assess at a later point whether we want to change the default.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

r-vdp committed:
nixos/installer/tools: refactor to allow overriding installer tools.

We turn the different tools into options so that they can be
enabled/disabled and so that the corresponding packages can be specified
or overridden.

jtojnar commented on nixos/modules/installer/tools/tools.nix:
```diff
@@ -10,119 +10,163 @@ let
     dir = "bin";
     isExecutable = true;
   });
```
Could you please split the refactoring into a separate commit? It is currently difficult to review the diff.

R-VdP replied:
@jtojnar done. I'm afraid the diff of the second commit is still quite large though, since things moved around quite a bit.

r-vdp committed:
nixos: rebuild less derivations on nixpkgs/config hash changes.

Introduce a top-level file nixos-version.json, and an option to have nixos-version
read this file from /run/current-system instead, of hard-coding all the values in the script.
When configured as such, this avoids rebuilding every derivation that depends on the system path
every time that the nixpkgs or configuration revision hash changes.
For my config, this change reduced the amount of derivations that need to be
rebuild after a trivial config change from 42 to 9.

R-VdP commented on nixos/modules/installer/tools/nixos-version.sh:
```diff
@@ -7,25 +7,25 @@ case "$1" in
     ;;
   --hash|--revision)
-    if ! [[ @revision@ =~ ^[0-9a-f]+$ ]]; then
+    revision="$(jq --raw-output 'if .nixpkgsRevision != null then .nixpkgsRevision else "null" end' < /run/current-system/nixos-version.json)"
```
Done, thanks!

SuperSandro2000 commented on nixos/modules/installer/tools/tools.nix:
```diff
@@ -228,19 +272,12 @@ in
-
-    system.build = {
-      inherit nixos-install nixos-generate-config nixos-option nixos-rebuild nixos-enter;
-    };
+      lib.attrValues (lib.mapAttrs (_: tool: tool.package) installerTools);
```
```suggestion
    environment.systemPackages = lib.attrValues (lib.mapAttrs (_: tool: tool.package) installerTools);
```
nit

SuperSandro2000 commented on nixos/modules/installer/tools/tools.nix:
```diff
@@ -228,19 +272,12 @@ in
-    };
+      lib.attrValues (lib.mapAttrs (_: tool: tool.package) installerTools);
+
+    documentation.man.man-db.skipPackages =
+      lib.mkIf (installerTools ? nixos-version) [ config.system.installerTools.nixos-version.package ];
```
```suggestion
    documentation.man.man-db.skipPackages = lib.mkIf (installerTools ? nixos-version) [ config.system.installerTools.nixos-version.package ];
```
nit

SuperSandro2000 commented on nixos/modules/installer/tools/tools.nix:
```diff
@@ -10,119 +10,163 @@ let
+      };
+
+      nixos-option = {
+        enable = lib.versionOlder (lib.getVersion config.nix.package) "2.4pre";
```
Good that I saw that, removed it in #237442 . Someone wants to review that?

R-VdP replied:
Hah, I looked for that PR because I remembered seeing that, but somehow it didn't show up in the search results. I think it's easy to rebase either of them on top of the other, depending on which one gets merged first.

I can look at the other suggestions you made in a bit.

SuperSandro2000 replied:
Feel free to merge this PR first, rebasing my change is not hard.

SuperSandro2000 commented on nixos/modules/installer/tools/tools.nix:
```diff
@@ -10,119 +10,163 @@ let
+            '';
+          };
+        };
+        config = {
+          name = lib.mkDefault name;
+        };
```
```suggestion
        config.name = lib.mkDefault name;
```
nit

SuperSandro2000 commented on nixos/modules/installer/tools/tools.nix:
```diff
@@ -228,19 +272,12 @@ in
-
-    system.build = {
-      inherit nixos-install nixos-generate-config nixos-option nixos-rebuild nixos-enter;
-    };
+      lib.attrValues (lib.mapAttrs (_: tool: tool.package) installerTools);
```
```suggestion
    environment.systemPackages = lib.attrValues (lib.mapAttrs (_: tool: tool.package) installerTools);
```
nit

R-VdP replied:
I think the line becomes very long like this, making it pretty annoying to read when using multiple panes in an editor.
It seems that there's nothing about line length in the editorconfig file, so not sure what's the convention here.

SuperSandro2000 reviewed (approved):
works 🎉 

![image](https://github.com/NixOS/nixpkgs/assets/7258858/fcf1cc22-f62a-4c16-a5b6-46278de20cba)

ajs124 commented on nixos/modules/installer/tools/nixos-version.sh:
```diff
@@ -7,25 +7,25 @@ case "$1" in
     ;;
   --hash|--revision)
-    if ! [[ @revision@ =~ ^[0-9a-f]+$ ]]; then
+    revision="$(jq --raw-output '.nixpkgsRevision // "null"' < /run/current-system/nixos-version.json)"
```
I feel like this has been proposed before and was decided against. Can try and look for previous PR/issue if you want me to.

The reasoning was that this doesn't do the same thing. Previously, nixos-version executed from a nixos system closure always gives you the version of that system.

I understand and share the desire to reduce rebuilding these derivations, but this should be carefully be decided one,  because it 100% breaks existing behavior.

cole-h replied:
Maybe this PR could just introduce the json file and a toggle for this script so that parties interested in getting rid of the constant rebuilds could remove the script, but still be able to get that information out by manually using `jq`.

emilazy replied:
A JSON file is probably easier to consume than a shell script for automated tooling that wants the exact version anyway, right? I realize this is a compatibility-breaking change but it seems worth it in the long run.

R-VdP replied:
What is the use-case for executing a script from a non-active nixos closure? Do you have any visibility on where this is used? I can think of deployment tools maybe.
I do also think though that parsing a top-level JSON file that has this data seems easier and safer than relying on the output of a script that you need to go look for. It's not very obvious either that that script exists if you didn't already know about it.

SuperSandro2000 replied:
> a toggle for this script so that parties interested in getting rid of the constant rebuilds could remove the script

we should really turn this on by default. Rebuilding and restarting those handful of configs is creating noise and sometimes even breaking things.

xaverdh replied:
So a good way forward would be to guard this with a configuration switch for the time being while additionally adding a `json` file to the configuration for any tooling that wants to parse it. Then we can inform potential users of the old way and switch the default later?

R-VdP replied:
That sounds reasonable to me. If we have an agreement that this is the way forward, then I can implement it this way.

emilazy replied:
Is there any way we can proactively surface this to tooling? Maybe print a warning to stderr when running `nixos-version` on an old version of the system?

emilazy replied:
I guess that's a little circular in that it involves comparing versions :) But "store path of `$0` != store path of `/run/current-system`" might be sufficient.

R-VdP replied:
I pushed a new version addressing these concerns. The new behaviour is behind an option now (`system.installerTools.nixos-version.selfContained`), which defaults to `true`, giving the old behaviour. In that case we substitute the store path of the JSON file directly into the `nixos-version` script (hence why it's "self-contained").

When setting this to false, `nixos-version` looks at `/run/current-system/nixos-version.json` instead, and so won't be changing with every change to the nixpkgs or config revision hashes. This behaviour must be explicitly enabled for now.

I'm open to suggestions for the names, if you think something else would be more suitable. I also wonder whether we'd at some point want to switch this default, or not.

R-VdP replied:
@ajs124 @xaverdh @emilazy do you think we can move forward after the changes I made to incorporate your feedback?

xaverdh replied:
I think this is fine now! We can still discuss whether to change defaults in the future, but that should not delay this pr.

emilazy replied:
Yeah I guess it would be best to wait for this to hit stable before introducing any kind of warning along the lines of what I suggested, since it could break tooling.

r-vdp committed:
release-notes: top-level nixos-version.json file and nixos-version option

R-VdP commented:
I added an entry to the release notes for this change.

xaverdh commented on nixos/modules/installer/tools/tools.nix:
```diff
@@ -10,238 +10,292 @@ let
+        };
+
+        nixos-option = {
+          enable = lib.versionOlder (lib.getVersion config.nix.package) "2.4pre";
```
Since https://github.com/NixOS/nixpkgs/pull/237442, this should just be `true`

R-VdP replied:
Yep, I rebased the PR!

xaverdh replied:
might be github, but to me it still looks like the old code before #237442 (which switched this to always on).

R-VdP replied:
Hmm, weird, I could have sworn that I made the change during the rebase. I fixed it now in any case, thanks for the heads up!

roberth commented on nixos/modules/system/activation/top-level.nix:
```diff
@@ -159,6 +160,33 @@ in
       };
     };
 
+    system.nixosVersionJsonData = lib.mkOption {
```
`Data` doesn't mean anything. I think it'd be fine to call this `nixosVersionJson` and have the json drv in a let binding instead of an option.

(Also I think it doesn't _have_ to be a separate drv, but that doesn't have to change either)

roberth commented on nixos/modules/installer/tools/tools.nix:
```diff
@@ -10,236 +10,292 @@ let
-        splice in the boot loader configuration.
+        Disable nixos-rebuild, nixos-generate-config, nixos-installer
+        and other NixOS tools. This is useful to shrink embedded,
+        read-only systems which are not expected to be rebuild or
```
```suggestion
        read-only systems which are not expected to be rebuilt or
```

roberth commented on nixos/modules/system/activation/top-level.nix:
```diff
@@ -159,6 +160,33 @@ in
+        nixpkgsRevision = config.system.nixos.revision;
+      } // optionalAttrs (config.system.configurationRevision != null) {
+        configurationRevision = config.system.configurationRevision;
+      };
```
Needs a `defaultText = lib.literalExample '' ... '';` for the manual.

roberth commented on nixos/modules/system/activation/top-level.nix:
```diff
@@ -159,6 +160,33 @@ in
+      } // optionalAttrs (config.system.nixos.codeName != null) {
+        nixosCodeName = config.system.nixos.codeName;
+      } // optionalAttrs (config.system.nixos.revision != null) {
+        nixpkgsRevision = config.system.nixos.revision;
```
Why is this a different name?

roberth reviewed (commented):
I think the extra untested code path is a problem. I'd prefer to switch everyone to the more efficient way.

I suppose one way that might break is if someone relies on `"$(/nix/var/nix/profiles/system-$n-link/sw/bin/nixos-version)"`.

Could we detect such an invocation and make it fail?
Otherwise it would give the wrong result.

Alternatively, you might be able to find the correct version based on `$0`.

roberth commented on nixos/modules/installer/tools/tools.nix:
```diff
@@ -10,236 +10,292 @@ let
+    in
+    lib.mkIf (config.nix.enable && !config.system.disableInstallerTools) {
+
+      system.installerTools.tools = {
+        nixos-build-vms = {
+          enable = true;
```
Why not `programs.nixos-build-vms`?
These are perfectly normal "user" commands that you might as well get from `nix shell` are whatever; no need to treat categorize them differently.

xaverdh commented:
Ok, so maybe a different approach that would also avoid rebuilding things that depend on the system path is to have one  `json` file per revision instead of a global one. That file will still have to rebuilt, but only that.

R-VdP commented:
> Ok, so maybe a different approach that would also avoid rebuilding things that depend on the system path is to have one `json` file per revision instead of a global one. That file will still have to rebuilt, but only that.

But the issue is that you need to find it, so then we'd be back at having a changing reference to the JSON file in the `nixos-version` script, and we'd be rebuilding the system path and all its dependents on every version bump again, right?

The idea here is to make it possible for `nixos-version` to find the JSON file without needing to rebuild the script itself, but that means that we are less fine-grained. Unless we can do something with `$0` instead, as suggested above.

xaverdh commented:
> > Ok, so maybe a different approach that would also avoid rebuilding things that depend on the system path is to have one `json` file per revision instead of a global one. That file will still have to rebuilt, but only that.
> 
> But the issue is that you need to find it, so then we'd be back at having a changing reference to the JSON file in the `nixos-version` script, and we'd be rebuilding the system path and all its dependents on every version bump again, right?
> 
> The idea here is to make it possible for `nixos-version` to find the JSON file without needing to rebuild the script itself, but that means that we are less fine-grained. Unless we can do something with `$0` instead, as suggested above.

Yes, I was thinking about using  `$0` to find the associated `json` file

R-VdP commented:
> > > Ok, so maybe a different approach that would also avoid rebuilding things that depend on the system path is to have one `json` file per revision instead of a global one. That file will still have to rebuilt, but only that.
> > 
> > 
> > But the issue is that you need to find it, so then we'd be back at having a changing reference to the JSON file in the `nixos-version` script, and we'd be rebuilding the system path and all its dependents on every version bump again, right?
> > The idea here is to make it possible for `nixos-version` to find the JSON file without needing to rebuild the script itself, but that means that we are less fine-grained. Unless we can do something with `$0` instead, as suggested above.
> 
> Yes, I was thinking of using `$0` to find the associated `json` file

I see three scenarios to consider:
1. Calling `/run/current-system/sw/bin/nixos-version`
1. Calling `/nix/var/nix/profiles/system-$n-link/sw/bin/nixos-version`
1. Calling `/nix/store/xxxxx-nixos-version/bin/nixos-version`
1. Calling `nixos-version` through some unknown custom symlink

We could do something like 
```
profileDir = "$(dirname $(dirname $(dirname $0)))"
versionInfo = "${profileDir}/nixos-version.json"

if [ ! -f "${versionInfo}" ]; then
  echo "..."
  exit 1
fi

<rest of the script>
```

Which would work for the first three. In the third case we throw an error, I don't think we have another option there.

The fourth one will depend on the structure of the symlink, but I think it's ok not to support this.

Do you see any other scenarios to consider?

R-VdP commented:
> why not just canonicalize the path with `realpath` before taking the directory? Then it shouldn't matter if we go through an additional indirection. I guess one can still come up with obscure cases where things break (eg. ppl messing with `$0` or copying `nixos-version` around without the profile), but that could be acceptable breakage in my opinion

You mean that we include the JSON file in the output of the nixos-version derivation but not have it be part of the system path?

Like have `/nix/store/xxxxxxx-nixos-version/share/nixos-version.json`, something like that.

In that case we could do something like `$(dirname $(dirname $(realpath $0)))/share/nixos-version.json` to get to the JSON file.

I would still like to have the JSON file in the top-level system closure though, and not depend on the `nixos-version` script necessarily. But I don't think we can reliably find this top-level dir from the `nixos-version` script, unless we assume that we're called from some link that allows us to `dirname` up to the profile dir.

roberth commented:
> * Calling `/run/current-system/sw/bin/nixos-version`
> * Calling `/nix/var/nix/profiles/system-$n-link/sw/bin/nixos-version`
> * Calling `/nix/store/xxxxx-nixos-version/bin/nixos-version`

I see test cases :star_struck:
